File: big-o.js
/*****************************************/
// Big O Notation examples in JavaScript //
/*****************************************/
// Run node big-o.js
/****************************************/
// Run 'var b = require('./big-o.js')'
// then 'var a = new b.bigONotation(100)'
// to create a new bigONotation object
/****************************************/
// Run a.bubbleSort()

function generateArray(size) {
  var newArray = [];

  for (var i=0; i < size; i++) {
    newArray.push(Math.floor(Math.random()*1000) + 10);
  }
  return newArray;
};

function swapValues(array, index1, index2) {
  var a = array[index1],
      b = array[index2];

  array[index2] = a;
  array[index1] = b;
};

var bigONotation = function(size) {
  this.array = generateArray(size);
  this.arraySize = size;
};

// O(1)
bigONotation.prototype.addToArray = function(val) {
  this.array.push(val);
};

// O(n) - linear search
bigONotation.prototype.linearSearch = function(value) {
  var startTime = new Date();
  for (var i=0; i < this.array.length; i++) {
    if (this.array[i] == value) {
      console.log('Found at index: ' + i); 
      console.log('Linear Search took ' + (new Date() - startTime)/1000 + 'ms');
      break;
    }
  }
};

// O(n^2) - bubble sort
bigONotation.prototype.bubbleSort = function() {
  var startTime = new Date();
  while (true) {
    var swapped = false;
    for (var i=0; i < this.array.length; i++) {
      if (i < this.array.length && this.array[i] > this.array[i+1]) {
        swapValues(this.array, i, i+1);
        swapped = true;
      }
    }
    if (!swapped) break;
  }
  console.log('Bubble sort took ' + (new Date() - startTime)/1000 + 'ms');
};

// O(log n) - binary search
bigONotation.prototype.binarySearch = function(value) {
  var startTime = new Date(),
      min = 0,
      max = this.arraySize - 1,
      iterations = 0;

  while (min <= max) {
    mid = Math.ceil((max + min)/2);
    if (this.array[mid] < value) {
      min = mid + 1;
    } else if (this.array[mid] > value) {
      max = mid - 1;
    } else {
      console.log('Found match in index '+ mid);
      min = max + 1;
    }
    iterations += 1;
  }
  console.log('Binary search took ' + (new Date() - startTime) + 'ms');
  console.log('Iterations: ' + iterations);
};

// O(n log n) - quick sort
bigONotation.prototype.quickSort = function(array) {
  if (array.length <= 1) return array;

  var pivotIndex = Math.ceil(array.length/2),
      pivotValue = array.splice(pivotIndex, 1)[0],
      left = [],
      right = [];

  for (var i=0; i < array.length; i++) {
    if (array[i] < pivotValue) {
      left.push(array[i]);
    } else {
      right.push(array[i]);
    }
  }
  return this.array = this.quickSort(left).concat([pivotValue], this.quickSort(right));
};

// quickSort + timer
bigONotation.prototype.callQuickSort = function(array) {
  var startTime = new Date(),
      resolvePromise = Promise.resolve(this.quickSort(array));

  resolvePromise.then(function(data) {
    console.log(data);
    console.log('Quick sort took ' + (new Date() - startTime)/1000 + 'ms');
  });
};

module.exports = {
  generateArray,
  bigONotation
}

File: big_o.rb
#################################
# Big O Notation examples in Ruby
#################################
# run irb -I . -r big_o.rb
#################################
# Example: 
# run 'a = BigONotation.new(10000)'
# then 'a.linear_search(20)'
# and compare to running 'b = BigONotation.new(20000)'
# and 'b.linear_search(20)
# the amount of time should increase proportionally with N (size of array) 
#################################

class BigONotation
  attr_accessor :items_in_array, :array

  def initialize(size)
    @array = Array.new(size)
    @array_size = size
    @items_in_array = 0
    generate_array(@array)
  end

  # O(1) - perform the same no matter the size of the array
  def add_to_array(new_item)
    start_time = Time.now
    @array[@items_in_array] = new_item
    end_time = Time.now
    @items_in_array += 1
  end

  # O(n) - performs in correlation to N
  def linear_search(value)
    start_time = Time.now.to_f
    
    @array.each do |a|
      if a == value
        end_time = Time.now.to_f
        return "Linear Search took #{(end_time - start_time).round(3)}s"  
      end
    end
  end

  # O(n^2) - performs to N squared (slow)
  def bubble_sort
    start_time = Time.now.to_f
    sort_size = @array_size - 1

    (sort_size).times do
      @array.each_with_index do |a,i|
        if i < sort_size && a > @array[i+1]
          swap_values(i,i+1)
        end
      end
    end

    end_time = Time.now.to_f
    p @array
    "Bubble Sort took #{(end_time - start_time).round(3)}s"  
  end

  # O(log n)
  def binary_search(value)
    start_time = Time.now.to_f
    low_index = 0
    high_index = @array_size - 1
    iterations = 0

    while low_index < high_index
      middle_index = (high_index + low_index)/2
      if @array[middle_index] < value
        low_index = middle_index + 1
      elsif @array[middle_index] > value
        high_index = middle_index - 1
      else 
        low_index = high_index + 1
        p "Found match in index #{middle_index}"
      end
      iterations += 1
    end

    end_time = Time.now.to_f
    p "Binary Search took #{(end_time - start_time).round(3)}s"  
    p "Iterations: #{iterations}"
  end

  # O(n log n)
  def quick_sort(array)
    return array if array.length <= 1

    pivot_index = (array.length / 2).to_i
    pivot_value = array[pivot_index]
    array.delete_at(pivot_index)
    left,right = array.partition{|x| x < pivot_value}
    return quick_sort(left) + [pivot_value] + quick_sort(right)
  end

  def call_quick_sort(array)
    start_time = Time.now.to_f
    end_array = quick_sort(array)
    end_time = Time.now.to_f
    p end_array
    p "Quick Sort took #{(end_time - start_time).round(6)}s"  
  end

  def generate_array(array)
    array.each_with_index do |a,i|
      array[i] = rand(1000)+10
    end
    @items_in_array = @array_size - 1
  end

  def swap_values(index1, index2)
    a = @array[index1]
    @array[index1] = @array[index2]
    @array[index2] = a
  end
end

